text/000-rocq-findlib.md

```diff
@@ -0,0 +1,267 @@
+libraries (including Rocq plugins), and one containing Rocq theories. One may
+run the following command to determine whether a package is a theory or not.
+```
+ocamlfind query some_dep -format "%(rocqpath)"
+```
+The result is empty if `some_dep` is a standard OCaml package (possibly a Rocq
+plugin), and it gives a directory path for Rocq theories.
+
+The mapping for each theory package `some_theory` can then be computed using
+the following command.
+```
+ocamlfind query some_theory -format "-Q %d/rocq.d %(rocqpath)"
+```
+
+## Example: stdpp
+
+The packages for [stdpp](https://gitlab.mpi-sws.org/iris/stdpp) could end up
+being installed under `<PREFIX>/lib/rocq-stdpp/`, which would contains the
+following `META` file.
+```
+version = "dev"
+description = "An extended \"Standard Library\" for Rocq."
+rocqpath = "stdpp"
+requires = "rocq-stdlib"
+
+package "bitvector" (
+  directory = "bitvector"
+  version = "dev"
+  description = "A library for bitvectors based on std++."
+  rocqpath = "stdpp.bitvector"
+  requires = "rocq-stdpp"
+)
+
+package "unstable" (
+  directory = "unstable"
+  version = "dev"
+  description = "Unfinished std++ libraries."
+  rocqpath = "stdpp.unstable"
+  requires = "rocq-stdpp"
+)
+```
+
+Differences with the current `stdpp` setup:
+- Rocq-ified package names.
+- Use of sub-packages form `stdpp.bitvector` and `stdpp.unstable`, instead of
+  different Opam packages.
+
+## Example: coq-elpi
+
+```
+rocqpath = "elpi"
+requires = "rocq-stdlib coq-elpi.plugin"
+...
+
+package "plugin" (
+  directory = "plugin"
+  requires = "rocq-core.plugins.ltac rocq-core.vernac elpi"
+  ...
+)
+
+package "apps" (
+  directory "apps"
+
+  package "derive" (
+    directory "derive"
+    rocqpath = "elpi.apps.derive"
+    requires = "rocq-elpi rocq-elpi.apps.derive.elpi"
+    ...
+
+    package "elpi" (
+      directory "elpi"
+      rocqpath = "elpi.apps.derive.elpi"
+      requires = "rocq-elpi"
+      ...
+    )
+  )
+
+  package "tc" (
+    directory "tc"
+    rocqpath = "elpi.apps.tc"
+    requires = "rocq-elpi rocq-elpi.apps.tc.elpi rocq-elpi.apps.tc.plugin"
+    ...
+
+    package "plugin" (
+      directory = "plugin"
+      requires = "rocq-core.plugins.ltac rocq-core.vernac rocq-elpi.plugin"
+      ...
+    )
+
+    package "elpi" (
+      directory "elpi"
+      rocqpath = "elpi.apps.tc.elpi"
+      requires = "rocq-elpi"
+      ...
+    )
+  )
+
+  ...
+)
+```
+
+# Drawbacks
+
+None, except the eventual need to port all projects to use a build system that
+respects the new installation scheme. However, as a first step, it is possible
+to keep using the current installation scheme at the same time.
+
+# Alternatives
+
+All reasonable systems have a modular installation scheme. An alternative
+implementation would be to start from scratch and not reuse the Findlib file
+format, but given the close integration of Rocq in the OCaml ecosystem that
+probably would not make sense.
+
+# Unresolved questions
```

Comment:
When dune is enhanced, perhaps we can also make progress on these related issues:
- From within Rocq, there's currently no way to learn what packages are installed.  The installation process should save the opam package name, version number, description and the associated Rocq load path names (-R -Q etc, determining the module name) so that Rocq can show this information.

- Rocq should be able to show package dependencies for installed packages.  Non-developer users won't want to wade through a bunch of configuration files to get this info.  (If they can get readily get it from the OS command line in an understandable format, that would be reasonable.)

- There is no mechanism to ensure that two packages don't use the same load path, in which case you can't use both simultaneously without modifying one of the packages.  This could be addressed with a registry of package load paths for Rocq opam packages and checking that new packages have unique load paths.  This would be an opam change rather than dune.

Reply:
I believe that most of these things could be done by querying `opam` and using `ocamlfind` (or the underlying `findlib` OCaml library). In any case, a tool could be developed to easily query that information if people find that is useful.

About your the last point: I believe there should be no such mechanism. It is perfectly fine if two independent packages use the same load path, even if two files end up a the same module path. This is even sometimes desirable, for example for configuration packages not meant to be used at the same time. If two packages define the same module, then the order of dependencies should be good enough for disambiguation.

Reply:
The doc section I mention above lists some useful opam commands.  I'm not aware of how you get the full list of dependencies for a package.  Also, there's no clean way to get the Coq load path (logical name) from opam.  Non-developer users do need to know that will want to know the dependencies.

For load paths, there could be a way to skip the check for the case you mention.

> If two packages define the same module, then the order of dependencies should be good enough for disambiguation.

Where does this order of dependencies come from?  You shouldn't have to specify -R on the command line for installed packages.  Also it would be strange to have a subdirectory that's in the overridden package that doesn't exist in the overriding package show up in the midst of the overriding package in `Print LoadPath`.  Don't you think that breaks the appearance of modularity.  Seems like this would be confusing for users.